src/cache.rs

```diff
@@ -0,0 +1,495 @@
+        if matches!(self.config, CacheConfig::None) {
+            return None;
+        }
+
+        self.transactions.get(hash)
+    }
+
+    /// Cache a transaction for the provided hash.
+    pub(crate) fn insert_transaction(&mut self, hash: H256, transaction: Transaction) {
+        if matches!(self.config, CacheConfig::None) {
+            return;
+        }
+
+        self.write_to_disk(
+            CACHE_TYPE_TRANSACTIONS,
+            format!("{:#x}", hash),
+            &transaction,
+        );
+        self.transactions.insert(hash, transaction);
+    }
+
+    /// Reads the cache contents from the disk, if available.
+    fn read_all_from_disk(&mut self, dir: &str) -> Result<(), String> {
+        for cache_type in [
+            CACHE_TYPE_BLOCKS_FULL,
+            CACHE_TYPE_BLOCKS_MIN,
+            CACHE_TYPE_BLOCK_RAW_TRANSACTIONS,
+            CACHE_TYPE_TRANSACTIONS,
+        ] {
+            let cache_dir = Path::new(dir).join(cache_type);
+            let dir_listing = fs::read_dir(cache_dir.clone())
+                .map_err(|err| format!("failed reading dir '{:?}': {:?}", cache_dir, err))?
+                .flatten();
+            for file in dir_listing {
+                let key = file
+                    .file_name()
+                    .to_str()
+                    .ok_or_else(|| String::from("failed converting filename to string"))?
+                    .to_string();
+
+                let cache_file = File::open(file.path()).map_err(|err| {
+                    format!("failed reading file: '{:?}': {:?}", file.path(), err)
+                })?;
+                let reader = BufReader::new(cache_file);
+                match cache_type {
+                    CACHE_TYPE_BLOCKS_FULL => {
+                        let key = H256::from_str(&key).map_err(|err| {
+                            format!("invalid key for cache file '{:?}': {:?}", key, err)
+                        })?;
+                        let block: Block<TransactionVariant> = serde_json::from_reader(reader)
+                            .map_err(|err| {
+                                format!("failed parsing json for cache file '{:?}': {:?}", key, err)
+                            })?;
+                        self.block_hashes.insert(block.number.as_u64(), block.hash);
+                        self.blocks_full.insert(key, block);
+                    }
+                    CACHE_TYPE_BLOCKS_MIN => {
+                        let key = H256::from_str(&key).map_err(|err| {
+                            format!("invalid key for cache file '{:?}': {:?}", key, err)
+                        })?;
+                        let block: Block<TransactionVariant> = serde_json::from_reader(reader)
+                            .map_err(|err| {
+                                format!("failed parsing json for cache file '{:?}': {:?}", key, err)
+                            })?;
+                        self.block_hashes.insert(block.number.as_u64(), block.hash);
```

Comment:
Is there any concern with re-using the same `block_hashes` for `blocks_min` and `blocks_full`? One scenario I'm worried about is:

1. Fork mainnet and cache the full block for `700_000`, which has a hash of `0x123`
2. Restart the node but fork testnet and cache the min block for `700_000`, which has a different hash of `0x456`
3. Restart the node and fork mainnet again
4. Calling `get_block_by_number` for `700_000` will first fetch the hash, which returns `0x456`. It will then attempt to fetch block `0x456` and fail (as it only exists on mainnet).

This scenario isn't too bad, as it's just a cache miss, but wondering if there's any other scenarios more nefarious of returning incorrect data. I _think_ that would only be possible if hashes collide, but let me know what you think

Reply:
I thought of the situation already and came to the following assessment (please add/comment where necessary)

A cache for `mainnet` and a cache for `testnet` would be incompatible with the current approach of only ever having a single cache location. To mitigate that, I thought of hashing the fork url as another indirection layer but opted not to include this as an overengineered solution. This is so, since I've assumed that in an average use case, the user might use a singular fork url to perform their operations and would not mind purging the cache when switching to a different network.

One can however argue, that we can support  fork url based caching anyway, so the user has a better UX and can use the test node without any surprises. In which case, I'd be happy to add the fork url based caching.

Reply:
I think the current approach is flexible enough with the ability to clear cache that I'm not concerned. We can always add the extra path of fork url to the cache, but for now it shouldn't block this PR.